### PR TITLE
CASEFLOW-108 #resolve Create CavcRemandProcessedLetterResponseWindowTask

### LINKED ISSUE #15309: Create new CavcRemandProcessedLetterResponseWindowTask type
## Description
As a Litigation Support user, when I complete a SendCavcRemandProcessedLetterTask, I want a new task to be created and placed on hold, so that I have something to indicate that I am awaiting a response from the appellant. 

## Acceptance criteria
- [ ] New task type CavcRemandProcessedLetterResponseWindowTask
- [ ]  When a SendCavcRemandProcessedLetterTask is completed, a CavcRemandProcessedLetterResponseWindowTask is created
   - [ ] As a child of the CavcTask
   - [ ] Assigned to the CavcLitigationSupportTeam
   - [ ] Is put on hold for 90 days by creating a child timed hold task
   - [ ] Give task action of "End hold early" while task is on hold
   - [ ] After 90 days, the task should come off hold and show up in the CAVC team's unassigned tab tab to be assigned/acted upon  

## ~Open~ Resolved questions
1. Should this automatically close? (If yes, put on hold ala [EvidenceSubmissionWindowTask by including TimeableTask and implementing when_timer_ends and timer_ends_at]( https://github.com/department-of-veterans-affairs/caseflow/blob/34b3fb4c0e676b05c8a824293e0666d5b2872735/app/models/tasks/evidence_submission_window_task.rb#L8-L18)) If so, what happens when the 90 day window expires (goes to distribution, new "no response" task, etc)? 
**NO**
1. Instead, should this pop back up as "assigned" and actionable after 90 days? (if yes, put on hold ala no show hearing tasks, where they are [automatically created with a timed hold task](https://github.com/department-of-veterans-affairs/caseflow/blob/7adbc4bdb0e06b0741e9576d0c682a4a40ae6b92/app/models/tasks/no_show_hearing_task.rb#L8-L18)
**YES**

## Background/context/resources
When a user completes a SendCavcRemandProcessedLetterTask to indicate that they have sent out the 90 day letter to the veteran notifying them that their remand has been processed, the veteran has 90 days to respond to the letter on how they would like to proceed. A CavcRemandProcessedLetterResponseWindowTask should be opened and put on hold for 90 days when a SendCavcRemandProcessedLetterTask is completed

Task actions will be added in #15324

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves #15309 (aka [CASEFLOW-108](https://vajira.max.gov/browse/CASEFLOW-108))

### Description
Create new CavcRemandProcessedLetterResponseWindowTask for 90-day duration with ability to "end hold early".

Also included the action to place task back on hold -- see  [Slack](https://dsva.slack.com/archives/C01BKL4CCRL/p1605913943002800?thread_ts=1605913086.002300&cid=C01BKL4CCRL).

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Verify when a SendCavcRemandProcessedLetterTask is completed, a CavcRemandProcessedLetterResponseWindowTask is created
   - [ ] As a child of the CavcTask
   - [ ] Assigned to the CavcLitigationSupportTeam
   - [ ] Is put on hold for 90 days by creating a child timed hold task
   - [ ] Give task action of "End hold early" while task is on hold
   - [ ] After 90 days, the task should come off hold and show up in the CAVC teams unassigned tab to be assigned/acted upon

### Testing Plan
Log in as CAVC_LIT_SUPPORT_ADMIN (VACO)
1. Go to the team view of Queue: http://localhost:3000/organizations/cavc-lit-support?tab=unassignedTab&page=1
1. Click on a case in Unassigned tab; note the case docket number
1. "Assign to person" and check that case is now under the "Assigned tab" in the team view (note the task type)
1. Check task tree in the Rails console

Log in as the assignee (in an incognito window)
1.  go to their queue: http://localhost:3000/queue?tab=assigned_person&page=1
1. Click on the assigned case and "Mark task complete"; not that case is in the Completed tab in the user's queue
1. Click on the case, note the user as the action to "End hold early". (You can test it now or later.)
1. Check task tree in the Rails console

Log in as CAVC_LIT_SUPPORT_ADMIN (VACO)
1. check that case is still under the "Assigned tab" in the team view (with the Window Task type)
1. Click on the case, note the user as the action to "End hold early". (You can test it now or later.)
1. Check task tree in the Rails console

Ending and restoring hold:
1. When you "End hold early", the case shows in the Unassigned tab  in the team view.
1. When you "Put task on hold", the case shows in the "Assigned tab" in the team view.
1. Check task tree in the Rails console

Hold ends due to time passing:
1. Travel 90+ days into the future
   ```ruby
   Timecop.travel(Time.zone.now + 90.days + 1.hour)
   TaskTimerJob.perform_now
   ```
1. The case shows in the Unassigned tab  in the team view.
1. Check task tree in the Rails console

### User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

Note window task instructions and "End hold early" action:
image
Case Timeline shows completed send letter task:
image

After clicking on "End hold early" action:
image
and after clicking Submit, user can "Put task on hold" (do we want this action to be available?):
image

### Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### CODE REVIEW

**Review by yoomlam — COMMENTED** [review]
Code commentary

File: spec/feature/queue/cavc_task_queue_spec.rb [hunk 1]
```diff
@@ -10,47 +10,113 @@
...
-
-      # Logged in as second user assignee (due to reassignment)
-      User.authenticate!(user: org_nonadmin2)
-      visit "queue/appeals/#{send_task.appeal.external_id}"
-
-      find(".cf-select__control", text: "Select an action").click
-      find("div", class: "cf-select__option", text: Constants.TASK_ACTIONS.MARK_COMPLETE.label).click
-      fill_in "completeTaskInstructions", with: "Letter sent."
-      click_on COPY::MARK_TASK_COMPLETE_BUTTON
-      expect(page).to have_content COPY::MARK_TASK_COMPLETE_CONFIRMATION % send_task.appeal.veteran_full_name
```
Comment (yoomlam) [inline spec/feature/queue/cavc_task_queue_spec.rb:53]: Decompose into `step`s below.

File: client/COPY.json [hunk 2]
```diff
@@ -372,6 +372,10 @@
   "CAVC_TASK_DEFAULT_INSTRUCTIONS": "This task will be auto-completed when all CAVC-related tasks have been closed.",
 
   "SEND_CAVC_REMAND_PROCESSED_LETTER_TASK_LABEL": "Send CAVC-Remand-Processed Letter Task",
+  "SEND_CAVC_REMAND_PROCESSED_LETTER_TASK_COMPLETE_MODAL_BODY": "Marking this task complete will put this case on hold for 90 days to allow the appellant to respond.",
```
Comment (yoomlam) [inline client/COPY.json:375]: Confirmed COPY with Geronimo [here](https://dsva.slack.com/archives/C01BKL4CCRL/p1605913170002400?thread_ts=1605913086.002300&cid=C01BKL4CCRL).

**Review by yoomlam — COMMENTED** [review]
More commentary

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -20,7 +23,6 @@ class TasksController < ApplicationController
     DocketSwitchRulingTask: DocketSwitchRulingTask,
     DocketSwitchDeniedTask: DocketSwitchDeniedTask,
     DocketSwitchGrantedTask: DocketSwitchGrantedTask,
-    EvidenceSubmissionWindowTask: EvidenceSubmissionWindowTask,
```
[the remark this answers is not in the document]
Comment (hschallhorn) [reply app/controllers/tasks_controller.rb:23]: hey, that's me!

File: app/models/tasks/cavc_remand_processed_letter_response_window_task.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      end
+    end
+  end
+
+  def self.label
+    COPY::CAVC_REMAND_PROCESSED_LETTER_RESP_WINDOW_TASK_LABEL
+  end
+
+  def default_instructions
+    [COPY::CAVC_TASK_DEFAULT_INSTRUCTIONS]
```
Comment (hschallhorn) [inline app/models/tasks/cavc_remand_processed_letter_response_window_task.rb:35]: should this be used as the default instructions above?

File: app/models/tasks/cavc_remand_processed_letter_response_window_task.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  USER_ACTIONS = [
+    Constants.TASK_ACTIONS.TOGGLE_TIMED_HOLD.to_h
+  ].freeze
+
+  def available_actions(_user)
+    USER_ACTIONS
```
Comment (hschallhorn) [inline app/models/tasks/cavc_remand_processed_letter_response_window_task.rb:43]: believe that if we do not check the user is a part of the organization here, we may be letting anyone in caseflow toggle the hold, will confirm in testing
```suggestion
  def available_actions(user)
    assigned_to.user_has_access?(user) ? USER_ACTIONS : []
```

File: spec/models/tasks/cavc_remand_processed_letter_response_window_task_spec.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      end
+
+      context "when user_task cannot be marked complete" do
+        before { allow(user_task).to receive(:update_from_params).and_raise(StandardError) }
+        it "does not create window_task" do
+          expect(user_task.available_actions(org_nonadmin)).to match_array SendCRPLetterTask::USER_ACTIONS
+          expect { window_task }.to raise_error(StandardError)
+          expect(user_task.appeal.tasks.where(type: CRPLRWindowTask.name).count).to eq 0
+          expect(user_task.status).to eq Constants.TASK_STATUSES.assigned
+        end
```
Comment (hschallhorn) [inline spec/models/tasks/cavc_remand_processed_letter_response_window_task_spec.rb:81]: Think testing behavior of what happens when a `SendCavcRemandProcessedLetterTask` is completed should go in that test file

File: spec/feature/queue/cavc_task_queue_spec.rb (hunk 1 above)
Comment (hschallhorn) [reply spec/feature/queue/cavc_task_queue_spec.rb:53]: 👌 Can't wait for the merge conflicts, but great idea!

File: spec/feature/queue/cavc_task_queue_spec.rb [hunk 7]
```diff
@@ -10,47 +10,113 @@
...
+
+        # Check that appeal is in correct tab in Team view
+        User.authenticate!(user: org_admin)
+        visit "organizations/cavc-lit-support"
+        find(".cf-tab", text: "Assigned").click
+        expect(page).to have_content send_task.appeal.docket_number
+        # Check that org_admin has option to End hold early
+        visit "queue/appeals/#{send_task.appeal.external_id}"
+        find(".cf-select__control", text: "Select an action").click
+        find("div", class: "cf-select__option", text: Constants.TASK_ACTIONS.END_TIMED_HOLD.label).click
```
Comment (hschallhorn) [inline spec/feature/queue/cavc_task_queue_spec.rb:71]: Helper function for all of these if you're interested!
```suggestion
        click_dropdown(text: Constants.TASK_ACTIONS.END_TIMED_HOLD.label)
```

File: app/repositories/task_action_repository.rb [hunk 8]
```diff
@@ -272,6 +272,8 @@ def complete_data(task, _user = nil)
                               COPY::NO_SHOW_HEARING_TASK_COMPLETE_MODAL_BODY
                             elsif task.is_a? HearingAdminActionTask
                               COPY::HEARING_SCHEDULE_COMPLETE_ADMIN_MODAL
+                            elsif task.is_a? SendCavcRemandProcessedLetterTask
+                              COPY::SEND_CAVC_REMAND_PROCESSED_LETTER_TASK_COMPLETE_MODAL_BODY
                             else
                               COPY::MARK_TASK_COMPLETE_COPY
```
Comment (hschallhorn) [inline app/repositories/task_action_repository.rb:278]: Can we break this out into a hash, mapping task type to the modal body?

File: app/models/tasks/cavc_remand_processed_letter_response_window_task.rb (hunk 5 above)
Comment (hschallhorn) [reply app/models/tasks/cavc_remand_processed_letter_response_window_task.rb:43]: Yep! This user is not on the cavc team and has the toggle hold action
Screen Shot 2020-11-23 at 3 49 17 PM

Comment (yoomlam) [reply app/models/tasks/cavc_remand_processed_letter_response_window_task.rb:43]: Also added test.

File: app/models/tasks/cavc_remand_processed_letter_response_window_task.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+  before_validation :set_assignee
+
+  def self.create_with_hold(parent_task)
+    multi_transaction do
+      create!(parent: parent_task, appeal: parent_task.appeal).tap do |window_task|
+        TimedHoldTask.create_from_parent(
+          window_task,
+          days_on_hold: 90,
+          instructions: ["This case is on hold for 90 days to allow the appellant to respond."]
```
Comment (hschallhorn) [inline app/models/tasks/cavc_remand_processed_letter_response_window_task.rb:24]: Last nit, let's just call default instructions here so we aren't repeating instructions and only have 1 place to update if we decide on new copy

File: spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb [hunk 10]
```diff
@@ -73,27 +78,42 @@
...
-        expect(send_task.available_actions(other_user)).to be_empty
+        expect(child_task.available_actions(other_user)).to be_empty
+      end
+    end
+
+    context "when SendCRPLetterTask completed" do
+      let(:user_task) { child_task }
+      subject { user_task.update_from_params({ status: Constants.TASK_STATUSES.completed }, org_nonadmin) }
+
+      it "status is updated to be completed" do
```
Comment (hschallhorn) [inline spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:105]: Do we also want to test that a CavcRemandProcessedLetterResponseWindowTask was automatically created on hold for 90 days?

Comment (yoomlam) [reply spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:105]: I'll add that check to `cavc_remand_processed_letter_response_window_task_spec.rb`.

File: spec/models/tasks/cavc_remand_processed_letter_response_window_task_spec.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        expect(new_task.assigned_to).to eq CavcLitigationSupport.singleton
+        expect(new_task.status).to eq Constants.TASK_STATUSES.on_hold
+
+        expect(appeal.tasks).to include new_task
+        expect(parent_task.children).to include new_task
+        child_timed_hold_tasks = new_task.children.where(type: :TimedHoldTask)
+        expect(child_timed_hold_tasks.count).to eq 1
+        expect(child_timed_hold_tasks.first.assigned_to).to eq CavcLitigationSupport.singleton
+        expect(child_timed_hold_tasks.first.status).to eq Constants.TASK_STATUSES.assigned
+        expect(child_timed_hold_tasks.first.timer_end_time.to_date).to eq(Time.zone.now.to_date + 90.days)
```
Comment (yoomlam) [inline spec/models/tasks/cavc_remand_processed_letter_response_window_task_spec.rb:39]: @hschallhorn For some reason:
```
child_timed_hold_tasks.first.days_on_hold
=> nil
```
so I'm checking the `timer_end_time` instead.

File: spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb (hunk 10 above)
Comment (hschallhorn) [reply spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:105]: I think confirming a response window is created upon completion of the send task should exist in the test for the send task

Comment (yoomlam) [reply spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:105]: Done.

**Review by hschallhorn — APPROVED** [review]
🙌 Nice work!

**Pull request merged**